content/tutorials/parallelization/SIMWE_parallelization.qmd

```diff
@@ -0,0 +1,804 @@
+Negative values from [r.sim.sediment](https://grass.osgeo.org/grass-devel/manuals/r.sim.sediment.html)
+output are treated as erosion and converted to sediment mass.
+Summary statistics are computed using [r.univar](https://grass.osgeo.org/grass-devel/manuals/r.univar.html).
+
+```{python}
+# Erosion: extract negative values and convert to positive mass [kg/m^2]
+tools.r_mapcalc(
+    f"erosion_{huc12} = if(erdep_{huc12} < 0, abs(erdep_{huc12}) * {simulation_time * 60}, null())"
+)
+erosion = tools.r_univar(map=f"erosion_{huc12}", format="json")
+```
+
+```text
+ 'null_cells': 418835,
+ 'cells': 599907,
+ 'min': 3.4630666334578564e-08,
+ 'max': 154011.390625,
+ 'range': 154011.39062496537,
+ 'mean': 98.83270611141184,
+ 'mean_of_abs': 98.83270611141184,
+ 'stddev': 1045.8673705761544,
+ 'variance': 1093838.556835879,
+ 'coeff_var': 1058.2199068769523,
+ 'sum': 17895835.761005566}
+```
+
+Deactivate the active raster mask, restoring operations to apply across the full region.
+
+```{python}
+mask.deactivate()
+```
+
+Display water depth (`depth`) and erosion-deposition (`erdep`) rasters for the selected subwatershed.
+
+```{python}
+basin_map = gj.InteractiveMap()
+basin_map.add_raster(f"depth_{huc12}")
+basin_map.add_raster(f"erdep_{huc12}")
+basin_map.show()
+```
+
+![](SIMWE_images/basin_030401010402_both.webp)
+
+# Parallelized hydrologic and sediment modeling
+
+To efficiently model erosion and deposition across a large number of HUC12 subwatersheds,
+we will use the workflow we just ran and create a
+script that uses Python's `multiprocessing` module to parallelize the workflow.
+Each subwatershed is processed independently in its own environment, which allows computations to run concurrently without interference.
+
+Each subwatershed needs to set different mask and computational region, however normally, those settings are global, and so for different mask and region for each parallel process, we will use mask and region context managers.
+
+* Masking is handled using `MaskManager`, a [context manager for setting and managing raster mask](https://grass.osgeo.org/grass-devel/manuals/libpython/grass.script.html#grass.script.MaskManager),
+making it possible to have custom mask for the current process. This feature is available only since GRASS 8.5.
+
+    ```python
+    with gs.MaskManager(mask_name=f"basin_{huc12}"):
+        # Run actual computation with active mask.
+        gs.run_command(...)
+    ```
+
+* Computational region is handled using `RegionManager`, a [context manager for setting and managing computational region](https://grass.osgeo.org/grass-devel/manuals/libpython/grass.script.html#grass.script.RegionManager),
+making it possible to have custom region for the current process. This feature is available only since GRASS 8.5.
+
+    ```python
+    with gs.RegionManager(vector=f"basin_{huc12}"):
+        # Run actual computation in the specified region.
+        gs.run_command(...)
+    ```
```

Comment:
While the previous text talks about both, mask is technically below region in the hierarchy, i.e., first comes region and the mask refines that selection, so I would go with region followed by mask by a list like this one.